cfgmgr/buffermgrdyn.cpp

```diff
@@ -0,0 +1,1620 @@
+#include <fstream>
+#include <iostream>
+#include <string.h>
+#include "logger.h"
+#include "dbconnector.h"
+#include "producerstatetable.h"
+#include "tokenize.h"
+#include "ipprefix.h"
+#include "timer.h"
+#include "buffermgrdyn.h"
+#include "bufferorch.h"
+#include "exec.h"
+#include "shellcmd.h"
+#include "schema.h"
+#include "warm_restart.h"
+
+/*
+ * Some Tips
+ * 1. All keys in this file are in format of APPL_DB key.
+ *    Key population:
+ *        On receiving item update from CONFIG_DB: key has been transformed into the format of APPL_DB
+ *        In intermal maps: table name removed from the index
+ * 2. Maintain maps for pools, profiles and PGs in CONFIG_DB and APPL_DB
+ * 3. Keys of maps in this file don't contain the TABLE_NAME
+ * 3. 
+ */
+using namespace std;
+using namespace swss;
+
+BufferMgrDynamic::BufferMgrDynamic(DBConnector *cfgDb, DBConnector *stateDb, DBConnector *applDb, const vector<TableConnector> &tables, shared_ptr<vector<KeyOpFieldsValuesTuple>> gearboxInfo = nullptr) :
+        Orch(tables),
+        m_applDb(applDb),
+        m_cfgPortTable(cfgDb, CFG_PORT_TABLE_NAME),
+        m_cfgCableLenTable(cfgDb, CFG_PORT_CABLE_LEN_TABLE_NAME),
+        m_cfgBufferProfileTable(cfgDb, CFG_BUFFER_PROFILE_TABLE_NAME),
+        m_cfgBufferPgTable(cfgDb, CFG_BUFFER_PG_TABLE_NAME),
+        m_cfgLosslessPgPoolTable(cfgDb, CFG_BUFFER_POOL_TABLE_NAME),
+        m_cfgDefaultLosslessBufferParam(cfgDb, CFG_DEFAULT_LOSSLESS_BUFFER_PARAMETER),
+        m_applBufferPoolTable(applDb, APP_BUFFER_POOL_TABLE_NAME),
+        m_applBufferProfileTable(applDb, APP_BUFFER_PROFILE_TABLE_NAME),
+        m_applBufferPgTable(applDb, APP_BUFFER_PG_TABLE_NAME),
+        m_applBufferQueueTable(applDb, APP_BUFFER_QUEUE_TABLE_NAME),
+        m_applBufferIngressProfileListTable(applDb, APP_BUFFER_PORT_INGRESS_PROFILE_LIST_NAME),
+        m_applBufferEgressProfileListTable(applDb, APP_BUFFER_PORT_EGRESS_PROFILE_LIST_NAME),
+        m_stateBufferMaximumTable(stateDb, STATE_BUFFER_MAXIMUM_VALUE_TABLE),
+        m_stateBufferPoolTable(stateDb, STATE_BUFFER_POOL_TABLE_NAME),
+        m_stateBufferProfileTable(stateDb, STATE_BUFFER_PROFILE_TABLE_NAME),
+        m_applPortTable(applDb, APP_PORT_TABLE_NAME),
+        m_portInitDone(false),
+        m_firstTimeCalculateBufferPool(true)
+{
+    SWSS_LOG_ENTER();
+
+    // Initialize the handler map
+    initTableHandlerMap();
+    parseGearboxInfo(gearboxInfo);
+
+    string platform = getenv("ASIC_VENDOR") ? getenv("ASIC_VENDOR") : "";
```

Comment:
These Json files are provided by vendors. They can do that at a per SKU/platform basis.
The Kia plugin are supposed to be provided on a per asic vendor basis, which is why the asic vendor is loaded here

Reply:
Vendors can provide different lua scripts, but plugin names(headroomPluginName,  bufferpoolPluginName  and checkHeadroomPluginName ) are per asic_vendor and not per SKU. How can this be different for different SKUs? Could you pls clarify?

Reply:
The lua plugin is provided on a per vendor  basis. Different platform/SKUs can have different parameters, like cell size. This can be reflected in two ways:
- parameter passed by buffermgrd when calling lua plugins
- the lua plugin can also fetch these parameters via reading the tables in database, like ASIC_TABLE in STATE DB for cell size.
Lua plug-ins are not provided for each platform/SKU.